031-custom_authentication.md

```diff
@@ -0,0 +1,198 @@
+listener.security.protocol.map=
+principal.builder.class=
+```
+
+- `<listener-name>` would be the name for a given listener. 
+- `<sasl-mechanism>` would be the sasl mechanism which you wish to configure. Seeing as users can configure/build their own sasl mechanisms, all values will be allowed for this field, rather than having an allowlist of the standard sasl mechanisms. 
+
+To walkthrough an actual example, let's try configuring custom authentication for an oauthbearer workflow, where we would like to set the following properties:
+
+```
+listener.name.<listener-name>.ssl.keystore.location=/tmp/kafka/cluster.keystore.p12
+listener.name.<listener-name>.ssl.keystore.password=<password>
+listener.name.<listener-name>.ssl.keystore.type=PKCS12
+listener.name.<listener-name>.oauthbearer.sasl.client.callback.handler.class=client.class
+listener.name.<listener-name>.oauthbearer.sasl.server.callback.handler.class=server.class
+listener.name.<listener-name>.oauthbearer.sasl.login.callback.handler.class=login.class
+listener.name.<listener-name>.oauthbearer.connections.max.reauth.ms=999999999
+listener.name.<listener-name>.sasl.enabled.mechanisms=oauthbearer
+listener.name.<listener-name>.oauthbearer.sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required ;
+principal.builder.class=SimplePrincipal.class
+listener.security.protocol.map=CONTROLPLANE-9090:SSL,REPLICATION-9091:SSL,TLS-9093:SASL_SSL
+```
+
+To achieve this, we'd simply add the following listener config to our Kafka custom resource.
+
+```
+    listeners:
+      - name: custom-auth-listener
+        port: 9093
+        type: internal
+        tls: true
+        configuration:
+          brokerCertChainAndKey:
+            secretName: my-secret
+            certificate: my-certifciate.crt
+            key: my-key.key
+        authentication:
+          type: custom
+          sasl: true
+          principal.builder.class: SimplePrincipal.class
+          listener-config:
+            oauthbearer.sasl.client.callback.handler.class: client.class
+            oauthbearer.sasl.server.callback.handler.class: server.class
+            oauthbearer.sasl.login.callback.handler.class: login.class
+            oauthbearer.connections.max.reauth.ms: 999999999
+            sasl.enabled.mechanisms: oauthbearer
+            oauthbearer.sasl.jaas.config: |
+              org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required ;
+          tlsTrustedCertificates: ...
+          secrets:
+            - name: example
+              mountPath: "/etc/example"
+```
+
+Then, when constructing the broker config, we’ll perform the following tasks:
+
+* `Principal`, if set, is set cluster-wide for all authentication methods. This is a limitation of Kafka, which only allows one principal to be specified for the entire cluster. If set, we need to ensure that no other listeners override this property, and if they do and are different, then fail-hard.
```

Comment:
As you side it's cluster-wide and the user could specify the `principal.builder.class` in the `spec.kafka.config` section as well ... why just don't leave it there? Otherwise we have to deal with which configuration takes precedence on the other one and as you said different listeners using custom authn could specify different principal builder class but operator should raise error.

Reply:
That's a fair point @ppatierno! The primary reason why I left `principal.builder.class` to be defined at the listener level rather than `spec.kafka.config` is that currently the OAuth listener defines it at such level (https://github.com/strimzi/strimzi-kafka-operator/blob/ef576db7975dc40b19b559464caf1ff5f1b70071/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java#L212). I also think this makes sense, as otherwise users would have to configure their OAuth configuration in two separate locations, rather than just one.

Reply:
My only concern is about the user making a mistake and setting it in the Kafka config section. The implementation should have a well known precedence order failing when there are such mistakes.

Reply:
Gotcha, that makes sense! Thankfully we already forbid users from setting the value in the Kafka config section: https://github.com/strimzi/strimzi-kafka-operator/blob/ef576db7975dc40b19b559464caf1ff5f1b70071/api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterSpec.java#L45

Reply:
Ok cool! I should have had a look to this before asking ;-) Thanks!